scipy/ndimage/_filters.py

```diff
@@ -1553,9 +1554,29 @@ def _rank_filter(input, rank, size=None, footprint=None, output=None,
             raise RuntimeError(
                 "A sequence of modes is not supported by non-separable rank "
                 "filters")
-        mode = _ni_support._extend_mode_to_code(mode)
-        _nd_image.rank_filter(input, rank, footprint, output, mode, cval,
-                              origins)
+        mode = _ni_support._extend_mode_to_code(mode, is_filter=True)
+        if input.ndim == 1:
+            if input.dtype in (np.int64, np.float64, np.float32):
+                x = input.copy()
+                x_out = output
+            elif input.dtype == np.float16:
+                x = input.astype('float32')
+                x_out = np.empty(x.shape, dtype='float32')
+            elif np.result_type(input, np.int64) == np.int64:
+                x = input.astype('int64')
+                x_out = np.empty(x.shape, dtype='int64')
+            elif input.dtype.kind in 'biu':
+                # cast any other boolean, integer or unsigned type to int64
+                x = input.astype('int64')
+                x_out = np.empty(x.shape, dtype='int64')
+            else:
+                raise RuntimeError('Unsupported array type')
+            cval = x.dtype.type(cval)
+            _rank_filter_1d.rank_filter(x, rank, footprint.size, x_out, mode, cval, origin)
+            if input.dtype not in (np.int64, np.float64, np.float32):
+                np.copyto(output, x_out, casting='unsafe')
```

Comment:
Now that I look closer, this `casting="unsafe"` does not look right. You've made sure that the input is of the right dtype that your workhorse routines can handle, good. Then this `copyto` kicks in for other dtypes --- is it because the `output` can be e.g. boolean and `x_out` is int64, so that the cast is actually safe but numpy does not know?

I'd expect that either the cast is safe (no loss of information) or if not, this PR breaks backwards compatibility (bool input and output used to work on main, and now it's lossy?).